docs/client-specifications/error-handling.md

```diff
@@ -1,3 +1,143 @@
+  At the same time, we need to ensure that the gRPC status information is retained as (non-descript, semi-hidden/private)
+  fields on the final error object, so that Momento engineers can more easily determine root cause when assisting customers.
+
+# Error Types
+
+SDKs must provide a `MomentoError` type, which should extend from the base error type for the language.  (Though we will
+not initially be throwing these error instances--only returning them--this allows us to add a "throwOnError" flag in the
+future if users express a preference for that behavior.  It also allows users to throw the objects themselves if they
+so choose, ensures we retain stack trace information, etc.)
+
+This type must expose at least the following properties:
+
+```typescript
+class MomentoError extends Error {
+  errorCode: MomentoErrorCode;
+  message: string;
+  transportDetails: MomentoErrorTransportDetails;
+}
+```
+
+Implementations may provide subtypes that map to individual error codes if idiomatic to the language.
+
+The `MomentoErrorCode` is an enum that be used to easily and concretely check error types in code.  Its values should
+include:
+
+```typescript
+enum MomentoErrorCode {
+  INVALID_ARGUMENT_ERROR = 'INVALID_ARGUMENT_ERROR',       // message: `Invalid argument passed to Momento client: ${detail}`
+  UNKNOWN_SERVICE_ERROR = 'UNKNOWN_SERVICE_ERROR',         // message: `The service returned an unknown response; please contact Momento: ${detail}`
+  ALREADY_EXISTS_ERROR = 'ALREADY_EXISTS_ERROR',           // message: `A cache with the specified name already exists.  To resolve this error, either delete the existing cache and make a new one, or use a different name.  Cache name: '${cacheName}'`
+  NOT_FOUND_ERROR = 'NOT_FOUND_ERROR',                     // message: `A cache with the specified name does not exist.  To resolve this error, make sure you have created the cache before attempting to use it.  Cache name; '${cacheName}'`
+  INTERNAL_SERVER_ERROR = 'INTERNAL_SERVER_ERROR',         // message: `An unexpected error occurred while trying to fulfill the request; please contact Momento: ${detail}`
+  PERMISSION_ERROR = 'PERMISSION_ERROR',                   // message: `Insufficient permissions to perform an operation on a cache: ${detail}`
+  AUTHENTICATION_ERROR = 'AUTHENTICATION_ERRORS',          // message: `Invalid authentication credentials to connect to cache service: ${detail}`
+  CANCELLED_ERROR = 'CANCELLED_ERROR',                     // message: `The request was cancelled by the server; please contact Momento: ${detail}`
+  LIMIT_EXCEEDED_ERROR = 'LIMIT_EXCEEDED_ERROR',           // message: `Request rate exceeded the limits for this account.  To resolve this error, reduce your request rate, or contact Momento to request a limit increase.`
+  BAD_REQUEST_ERROR = 'BAD_REQUEST_ERROR',                 // message: `The request was invalid; please contact Momento: ${detail}`
```

Comment:
For some of these it would be rad to be able to provide better detail - like "The request was invalid; maximum request size (key + value) is 123456 bytes. Attempted 234567."

Reply:
For the specification, I included `${detail}` to try to capture places where the implementations should add more info if they have it.  For 'bad request' in particular, it seems like it would be challenging to represent all of the possible detail messages in the specification.

That said - if you have suggestions for how I can elaborate on the `${detail}` bit in the spec here to make it more clear that the implementations should provide as much detail as possible, lmk.